### PR TITLE
feat: add frontend for download performance

### PR DESCRIPTION
**Pull request by thaaaarun**
Fixes #<issue_number_goes_here>

> It's a good idea to open an issue first for discussion.

- [ ] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,3 @@
-# Cloud Storage Regionalized Performance
\ No newline at end of file
+# Cloud Storage Regionalized Performance
+This is a tool that allows you to measure download & upload performance
+between your location and other GCS bucket locations.
```
Comment (shaffeeullah) [inline README.md:3]: nit: change `your location` to `your client`

File: front_end/package.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{
```
Comment (shaffeeullah) [inline front_end/package.json:1]: won't let me comment on `front_end/package-lock.json`, but can you remove that file and add it to the .gitignore?

File: front_end/src/backend/downloads.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
+// In these functions the 'fileName' parameter variables refer to filenames of '2mib.txt' '64mib.txt' and '256mib.txt'
```
Comment (shaffeeullah) [inline front_end/src/backend/downloads.js:1]: thanks for documenting this! :)

File: front_end/src/backend/downloads.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    const bucketName = 'gcsrbpa-' + bucket;
+
+    // Build the URL - example: https://storage.googleapis.com/gcsrbpa-asia-southeast2/256mib.txt
+    // `https://storage.googleapis.com/${bucketName}/${fileName}`
+
+    const URL = `https://storage.googleapis.com/${bucketName}/${fileName}`;
+
+    // HTTP Request with Axios
+    async function downloadFile() {
+        // console.log(URL);
```
Comment (shaffeeullah) [inline front_end/src/backend/downloads.js:52]: comment can be removed

File: front_end/src/backend/downloads.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    const timeTaken = await downloadFile();
+    return timeTaken / 1000; // return in units of seconds
+}
+
+async function benchmark_single_download(fileName, bucketName, regions_json) {
+    var result = new Map();
+    const timeTaken = await timed_download(fileName, bucketName);
+
+    // Add one for throughput or bytes/second
+    var fileSizeBytes = 0;
```
Comment (shaffeeullah) [inline front_end/src/backend/downloads.js:68]: use `let` instead of `var`

File: front_end/tests/unit/example.spec.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import { shallowMount } from '@vue/test-utils'
```
Comment (shaffeeullah) [inline front_end/tests/unit/example.spec.js:1]: is this file needed?

File: front_end/src/views/HomeView.vue [hunk 7]
```diff
@@ -0,0 +1,18 @@
+<!-- <template>
```
Comment (shaffeeullah) [inline front_end/src/views/HomeView.vue:1]: is this file needed?

**Review by shaffeeullah — COMMENTED** [review]
This is awesome!! Left a few small comments after taking a quick first pass. @ddelgrosso1  can you please review as well?

**Comment by shaffeeullah** [comment]
@thaaaarun can you also lint these files? I should have github checks set up for that and the conventional commits. I'll figure out how to do that today

File: front_end/.browserslistrc [hunk 8]
```diff
@@ -0,0 +1,4 @@
+> 1%
```
Comment (ddelgrosso1) [inline front_end/.browserslistrc:1]: I'm not sure how important this is but you may want to check what Google policy is regarding supported browsers (if you didn't of course).

File: front_end/src/backend/downloads.js [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    "asia-south2":"Delhi",
+    "asia-southeast1":"Singapore",
+    "asia-southeast2":"Jakarta",
+    "australia-southeast1":"Sydney",
+    "australia-southeast2":"Melbourne"
+}`;
+let regions_json = JSON.parse(raw_data);
+
+
+async function timed_download(fileName, bucket) {
```
Comment (ddelgrosso1) [inline front_end/src/backend/downloads.js:41]: I believe our standard is to use `camelCase` as opposed to `snake_case` for JS/TS.

File: front_end/src/backend/downloads.js [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    const allBucketsResults = await benchmark_all_downloads(fileName);
+
+    if (log == 'log') {
+        console.log(allBucketsResults);
+    }
+
+    return JSON.stringify(allBucketsResults);
+}
+
+exports.timed_download = timed_download;
```
Comment (ddelgrosso1) [inline front_end/src/backend/downloads.js:134]: I believe you can put these in the function signatures as opposed to here, i.e. `export async function fooBar()`.

File: front_end/src/views/AboutView.vue [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<template>
```
Comment (ddelgrosso1) [inline front_end/src/views/AboutView.vue:1]: If this is just boiler plate from the generator and not going to be used it can be removed from the commit.

File: front_end/src/views/DownloadsView.vue [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      const fileName = currentFileSize + ".txt";
+
+      for (var bucketName in regions_json) {
+        try {
+          var result = await downloads.benchmark_download(fileName, bucketName);
+          this.results = this.results.concat(JSON.parse(result));
+          this.results = this.results.sort((a, b) => {
+            return a.timeTaken < b.timeTaken
+              ? -1
+              : a.timeTaken > b.timeTaken
```
Comment (ddelgrosso1) [inline front_end/src/views/DownloadsView.vue:48]: I would avoid nesting the ternary operator like this as it makes it a bit hard to read.

**Comment by ddelgrosso1** [comment]
I think you can remove the default favicon as well.

File: front_end/package.json [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    "@vue/cli-plugin-eslint": "~5.0.0",
+    "@vue/cli-plugin-router": "~5.0.0",
+    "@vue/cli-plugin-unit-jest": "~5.0.0",
+    "@vue/cli-service": "~5.0.0",
+    "@vue/test-utils": "^2.0.0-0",
+    "@vue/vue3-jest": "^27.0.0-alpha.1",
+    "babel-jest": "^27.0.6",
+    "eslint": "^7.32.0",
+    "eslint-plugin-vue": "^8.0.3",
+    "jest": "^27.0.5",
```
[the remark this answers is not in the document]
Comment (thaaaarun) [reply front_end/package.json:31]: There are no unit tests yet, we can remove jest for now and file this under issue #5

File: front_end/README.md [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+```
+
+### Compiles and hot-reloads for development
+```
+npm run serve
+```
+
+### Compiles and minifies for production
+```
+npm run build
```
Comment (ddelgrosso1) [inline front_end/README.md:15]: Just for consistency with most of our other libraries you might update this to `npm run compile`

File: front_end/README.md [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+```
+
+### Run your unit tests
+```
+npm run test:unit
```
Comment (ddelgrosso1) [inline front_end/README.md:20]: As above I would probably make this `npm run test` for consistency.

File: front_end/.browserslistrc (hunk 8 above)
Comment (thaaaarun) [reply front_end/.browserslistrc:1]: I did some research on this but couldn't come to a conclusion. Created and issue for this: #12

File: front_end/src/backend/downloads.js [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+    "asia-northeast2":"Osaka",
+    "asia-northeast3":"Seoul",
+    "asia-south1":"Mumbai",
+    "asia-south2":"Delhi",
+    "asia-southeast1":"Singapore",
+    "asia-southeast2":"Jakarta",
+    "australia-southeast1":"Sydney",
+    "australia-southeast2":"Melbourne"
+}`;
+const REGIONS_MAP = JSON.parse(REGIONS_JSON_STRING);
```
Comment (ddelgrosso1) [inline front_end/src/backend/downloads.js:54]: Is there a reason to treat the regions as a JSON string as opposed to just creating an object and avoiding the parse operation here?

Comment (ddelgrosso1) [reply front_end/src/backend/downloads.js]: I would just create an object here and avoid the extraneous call to convert (JSON.parse) from a string which appears unneeded in this context.

**Comment by ddelgrosso1** [comment]
One last comment, you might want to move the package.json and .gitignore out of the frontend folder to the root level. Overall, things look good and I ran it successfully in my cloudtop environment.

File: front_end/src/backend/downloads.js [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+
+    return bucketResults;
+}
+
+export async function benchmarkDownload(fileName, bucketName, log) {
+    const result = await benchmarkSingleDownload(fileName, bucketName, REGIONS_MAP);
+
+    console.log(`Completed Downloads Benchmark for ${bucketName}`);
+
+    if (log == 'log') {
```
Comment (JesseLovelace) [inline front_end/src/backend/downloads.js:127]: I'm a little confused about this, you need to pass in the string 'log' to log? Why? Just to avoid too much on the console log?

File: front_end/src/views/DownloadsView.vue [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+      //reset progress bar
+      progressBarCount = 0;
+      this.progressBarWidth = progressBarCount.toString() + "%";
+
+      this.results = [];
+      const fileName = currentFileSize + ".txt";
+
+      for (let bucketName in REGIONS_MAP) {
+        try {
+          let result = await downloads.benchmarkDownload(fileName, bucketName);
```
Comment (JesseLovelace) [inline front_end/src/views/DownloadsView.vue:59]: Doesn't this call the signature `benchmarkDownload(filename, log)`, meaning bucketName here would actually be log? Also, that function calls `benchmarkAllDownloads`, which also iterates through all of REGIONS_MAP, so doesn't this iterate through regions map twice?

Comment (thaaaarun) [reply front_end/src/views/DownloadsView.vue:59]: This calls `benchmarkDownload(fileName, bucketName, log)` so bucketName wouldn't be treated as the log parameter. 
`benchmarkDownload` also does not call `benchmarkAllDownloads`, it calls `benchmarkSingleDownload(fileName, bucketName, REGIONS_MAP)`. 

There are two similarly named functions `benchmarkDownload` and `benchmarkDownloads`

Comment (JesseLovelace) [reply front_end/src/views/DownloadsView.vue:59]: Ah yup i mixed them up because of the similar names

**Review by ddelgrosso1 — APPROVED** [review]
